File: omnigent/work_lifecycle.py
"""Shared Task, TaskRun, and Inbox lifecycle contracts.

These models define the Phase 0 wire contract only. Persistence and delivery
policy belong to the Task and Inbox services introduced in Phase 2.
"""

from __future__ import annotations

from enum import StrEnum
from typing import Any

from pydantic import AwareDatetime, BaseModel, ConfigDict, Field, model_validator


class TaskState(StrEnum):
    """User-facing Task workflow states."""

    BACKLOG = "backlog"
    TODO = "todo"
    IN_PROGRESS = "in_progress"
    REVIEW = "review"
    DONE = "done"
    BLOCKED = "blocked"
    FAILED = "failed"
    CANCELLED = "cancelled"


class TaskRunState(StrEnum):
    """Execution states for one TaskRun."""

    QUEUED = "queued"
    RUNNING = "running"
    WAITING = "waiting"
    SUCCEEDED = "succeeded"
    FAILED = "failed"
    CANCELLED = "cancelled"


class TaskRunTrigger(StrEnum):
    """Why a TaskRun was created."""

    MANUAL = "manual"
    RETRY = "retry"


class TaskLifecycleEventType(StrEnum):
    """Stable lifecycle event names shared by API and clients."""

    TASK_CREATED = "task.created"
    TASK_UPDATED = "task.updated"
    TASK_COMPLETED = "task.completed"
    TASK_CANCELLED = "task.cancelled"
    RUN_QUEUED = "run.queued"
    RUN_STARTED = "run.started"
    RUN_PROGRESS = "run.progress"
    RUN_WAITING = "run.waiting"
    RUN_SUCCEEDED = "run.succeeded"
    RUN_FAILED = "run.failed"
    RUN_CANCELLED = "run.cancelled"


class InboxCategory(StrEnum):
    """User-facing Inbox projection categories."""

    ACTION_REQUIRED = "action_required"
    PROGRESS = "progress"
    COMPLETED = "completed"
    FAILED = "failed"


class LifecycleEventSource(BaseModel):
    """Actor or subsystem that produced a lifecycle event."""

    kind: str = Field(min_length=1, max_length=64)
    id: str | None = Field(default=None, min_length=1, max_length=256)
    display_name: str | None = Field(default=None, min_length=1, max_length=256)

    model_config = ConfigDict(extra="forbid")


class RequiredAction(BaseModel):
    """Action metadata attached to a waiting event or Inbox item."""

    action_id: str = Field(min_length=1, max_length=256)
    kind: str = Field(min_length=1, max_length=64)
    prompt: str = Field(min_length=1, max_length=4000)
    expires_at: AwareDatetime | None = None

    model_config = ConfigDict(extra="forbid")


class FailureDetails(BaseModel):
    """Structured failure information safe to show in management views."""

    code: str = Field(min_length=1, max_length=128)
    message: str = Field(min_length=1, max_length=4000)
    retryable: bool = False

    model_config = ConfigDict(extra="forbid")


class ResultDetails(BaseModel):
    """Compact result metadata; artifacts remain referenced, not embedded."""

    summary: str | None = Field(default=None, max_length=4000)
    artifact_ids: list[str] = Field(default_factory=list)

    model_config = ConfigDict(extra="forbid")


class TaskLifecycleEvent(BaseModel):
    """Normalized immutable fact emitted by Task and TaskRun workflows."""

    event_id: str = Field(min_length=1, max_length=256)
    idempotency_key: str = Field(min_length=1, max_length=512)
    type: TaskLifecycleEventType
    occurred_at: AwareDatetime
    task_id: str = Field(min_length=1, max_length=256)
    run_id: str | None = Field(default=None, min_length=1, max_length=256)
    session_id: str | None = Field(default=None, min_length=1, max_length=256)
    project_id: str | None = Field(default=None, min_length=1, max_length=256)
    source: LifecycleEventSource
    summary: str = Field(min_length=1, max_length=1000)
    task_state: TaskState | None = None
    run_state: TaskRunState | None = None
    action: RequiredAction | None = None
    failure: FailureDetails | None = None
    result: ResultDetails | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)

    model_config = ConfigDict(extra="forbid")

    @model_validator(mode="after")
    def validate_event_shape(self) -> TaskLifecycleEvent:
        """Require execution identity and failure details when relevant."""
        if self.type.value.startswith("run.") and self.run_id is None:
            raise ValueError("run lifecycle events require run_id")
        if self.type is TaskLifecycleEventType.RUN_FAILED and self.failure is None:
            raise ValueError("run.failed events require failure details")
        return self


class InboxItemProjection(BaseModel):
    """Persistent Inbox view projected from one lifecycle event."""

    item_id: str = Field(min_length=1, max_length=256)
    dedupe_key: str = Field(min_length=1, max_length=512)
    source_event_id: str = Field(min_length=1, max_length=256)
    category: InboxCategory
    task_id: str = Field(min_length=1, max_length=256)
    run_id: str | None = Field(default=None, min_length=1, max_length=256)
    session_id: str | None = Field(default=None, min_length=1, max_length=256)
    project_id: str | None = Field(default=None, min_length=1, max_length=256)
    title: str = Field(min_length=1, max_length=256)
    summary: str = Field(min_length=1, max_length=1000)
    action: RequiredAction | None = None
    is_unread: bool
    created_at: AwareDatetime

    model_config = ConfigDict(extra="forbid")

File: web/src/lib/workLifecycle.ts
import { z } from "zod";

export const TASK_STATES = [
  "backlog",
  "todo",
  "in_progress",
  "review",
  "done",
  "cancelled",
] as const;

export const TASK_RUN_STATES = [
  "queued",
  "running",
  "waiting",
  "succeeded",
  "failed",
  "cancelled",
] as const;

export const TASK_LIFECYCLE_EVENT_TYPES = [
  "task.created",
  "task.updated",
  "task.completed",
  "task.cancelled",
  "run.queued",
  "run.started",
  "run.progress",
  "run.waiting",
  "run.succeeded",
  "run.failed",
  "run.cancelled",
] as const;

export const INBOX_CATEGORIES = ["action_required", "progress", "completed", "failed"] as const;

const nonEmptyId = z.string().min(1).max(256);
const awareIsoTimestamp = z
  .string()
  .refine(
    (value) => /(?:Z|[+-]\d{2}:\d{2})$/.test(value) && !Number.isNaN(Date.parse(value)),
    "Expected an ISO 8601 timestamp with timezone",
  );

export const lifecycleEventSourceSchema = z
  .object({
    kind: z.string().min(1).max(64),
    id: nonEmptyId.optional(),
    display_name: nonEmptyId.optional(),
  })
  .strict();

export const requiredActionSchema = z
  .object({
    action_id: nonEmptyId,
    kind: z.string().min(1).max(64),
    prompt: z.string().min(1).max(4000),
    expires_at: awareIsoTimestamp.optional(),
  })
  .strict();

export const failureDetailsSchema = z
  .object({
    code: z.string().min(1).max(128),
    message: z.string().min(1).max(4000),
    retryable: z.boolean().default(false),
  })
  .strict();

export const resultDetailsSchema = z
  .object({
    summary: z.string().max(4000).optional(),
    artifact_ids: z.array(z.string().min(1)).default([]),
  })
  .strict();

export const taskLifecycleEventSchema = z
  .object({
    event_id: nonEmptyId,
    idempotency_key: z.string().min(1).max(512),
    type: z.enum(TASK_LIFECYCLE_EVENT_TYPES),
    occurred_at: awareIsoTimestamp,
    task_id: nonEmptyId,
    run_id: nonEmptyId.optional(),
    session_id: nonEmptyId.optional(),
    project_id: nonEmptyId.optional(),
    source: lifecycleEventSourceSchema,
    summary: z.string().min(1).max(1000),
    task_state: z.enum(TASK_STATES).optional(),
    run_state: z.enum(TASK_RUN_STATES).optional(),
    action: requiredActionSchema.optional(),
    failure: failureDetailsSchema.optional(),
    result: resultDetailsSchema.optional(),
    metadata: z.record(z.string(), z.unknown()).default({}),
  })
  .strict()
  .superRefine((event, context) => {
    if (event.type.startsWith("run.") && event.run_id === undefined) {
      context.addIssue({ code: "custom", path: ["run_id"], message: "Run events require run_id" });
    }
    if (event.type === "run.failed" && event.failure === undefined) {
      context.addIssue({
        code: "custom",
        path: ["failure"],
        message: "run.failed events require failure details",
      });
    }
  });

export const inboxItemProjectionSchema = z
  .object({
    item_id: nonEmptyId,
    dedupe_key: z.string().min(1).max(512),
    source_event_id: nonEmptyId,
    category: z.enum(INBOX_CATEGORIES),
    task_id: nonEmptyId,
    run_id: nonEmptyId.optional(),
    session_id: nonEmptyId.optional(),
    project_id: nonEmptyId.optional(),
    title: z.string().min(1).max(256),
    summary: z.string().min(1).max(1000),
    action: requiredActionSchema.optional(),
    is_unread: z.boolean(),
    created_at: awareIsoTimestamp,
  })
  .strict();

export type TaskState = (typeof TASK_STATES)[number];
export type TaskRunState = (typeof TASK_RUN_STATES)[number];
export type TaskLifecycleEvent = z.infer<typeof taskLifecycleEventSchema>;
export type InboxItemProjection = z.infer<typeof inboxItemProjectionSchema>;

export function parseTaskLifecycleEvent(input: unknown): TaskLifecycleEvent {
  return taskLifecycleEventSchema.parse(input);
}

export function parseInboxItemProjection(input: unknown): InboxItemProjection {
  return inboxItemProjectionSchema.parse(input);
}
